library/common/event/provisional_dispatcher.h

```diff
@@ -0,0 +1,67 @@
+#pragma once
+
+#include "envoy/event/deferred_deletable.h"
+#include "envoy/event/dispatcher.h"
+
+#include "common/common/logger.h"
+#include "common/common/thread.h"
+#include "common/common/thread_synchronizer.h"
+
+#include "library/common/types/c_types.h"
+
+namespace Envoy {
+namespace Event {
+
+/**
+ * Wrapper around Envoy's Event::Dispatcher that queues callbacks until drain() is called. Future
+ * versions may support correct calling semantics after the underlying dispatcher has been
+ * terminated/deleted or before it has been created.
+ */
+class ProvisionalDispatcher : public Logger::Loggable<Logger::Id::main> {
+public:
+  ProvisionalDispatcher(Event::Dispatcher& event_dispatcher)
+      : event_dispatcher_(event_dispatcher) {}
+
+  /**
+   * Drains all queued callbacks to the real dispatcher. Must be called after the underlying
+   * dispatcher is running. Further posts will be transparently passed through.
+   */
+  void drain();
+
+  // TODO(goaway): return ENVOY_FAILURE after the underlying dispatcher has exited.
+  /**
+   * Before the underlying dispatcher is running, queues posted callbacks; afterwards passes them
+   * through.
+   * @param callback, the callback to be dispatched.
+   * @return should return ENVOY_FAILURE when the underlying dispatcher exits, but at present it
+   * always returns ENVOY_SUCCESS.
+   */
+  envoy_status_t post(Event::PostCb callback);
+
+  /**
+   * @return false before the underlying dispatcher is running, otherwise the result of the
+   * underlying call to Event::Dispatcher::isThreadSafe().
+   */
+  bool isThreadSafe();
+
+  /**
+   * Submits an item for deferred delete. Must be called from context where
+   * ProvisionalDispatcher::isThreadSafe() is true.
+   */
+  void deferredDelete(DeferredDeletablePtr&& to_delete);
+
+  // Used for testing.
+  Thread::ThreadSynchronizer& synchronizer() { return synchronizer_; }
+
+private:
+  // TODO(goaway): This class supports a straightforward case-specific lock-free implementation, but
```

Comment:
Can we document this in an issue. 

Is this referring to what we discussed a couple weeks ago? I thought we were going to go for the lock-free approach in this PR.

Reply:
Sorry for the confusion.

This PR isn't meant to remove all locks. It removes all locks from the HTTP client and simplifies reasoning about concurrency. Somewhat less locking will occur on the golden path.

This comment refers to a specialized, lock-free data structure that could be implemented here, based on the specific usage of the queue (many writers, one reader, one full non-atomic drain pass). It's an optimization that is straightforward enough to realize, but not a priority over the overall restructuring of the code.